### DISCUSSION TITLE
尝试并入上游

### DISCUSSION THREAD

**Discussion by easyteacher** [BLumia/pineapple-calendar#7] [Plasmoid / Applet Development]
整体思路参考了作者的文档，农历部分参考了这个项目的代码，非常感谢作者让我少走了很多弯路，也希望作者能帮忙审阅这几个合并请求，提高并入上游的可能性。

涉及了[KDeclarative](https://invent.kde.org/frameworks/kdeclarative)、[KDE Plasma Framework](https://invent.kde.org/frameworks/plasma-framework)和[Plasma Add-ons](https://invent.kde.org/plasma/kdeplasma-addons)三个项目。

1. https://invent.kde.org/frameworks/kdeclarative/-/merge_requests/112
2. https://invent.kde.org/frameworks/plasma-framework/-/merge_requests/450
3. https://invent.kde.org/plasma/kdeplasma-addons/-/merge_requests/112

目前的问题：

1. 想在标题栏上显示农历日期，还需要对plasma-workspace的[数字时钟插件](https://github.com/KDE/plasma-workspace/blob/master/applets/digital-clock/package/contents/ui/CalendarView.qml)进行修改，以及从MonthView里暴露出相关属性以避免重复建立DaysModel。
2. SubLabel还可以用于显示节假日，可能涉及到KHolidays项目。目前SubLabel里的yearLabel和monthLabel还没有实际用途，因为月份和年份的数据模型是ListModel，但感觉单独为月份和年份新建数据模型有点浪费（目前没有看到有日历需要在年份或月份下显示文字的）。
3. 不确定在Plasma里依赖ICU是否可行（虽然是可选的，且ICU一般发行版都有预装）

**Comment by BLumia** [comment]
早些时候看到了 bugtracker 里对应的 comment，工作量确实不小，辛苦了！

KDE 这边似乎本身也有对日历标签的支持意愿，之前试图推进这个功能到上游但很久没相关动静后，kde-pim 邮件列表里突然有重新提到过一次这件事 [1](https://mail.kde.org/pipermail/kde-pim/2021-October/048111.html) [2](https://mail.kde.org/pipermail/kde-pim/2021-November/048137.html)（不过似乎因为我没订阅 kde-pim 列表所以我回的邮件似乎没成功发到邮件列表里）。大概是希望加通用一些的 calendar decoration 的支持，以便用在 applet 之外的地方，比如 Kalendar 之类（倒是不知道为什么那边没提到 [calindori](https://github.com/KDE/calindori)）。

我原本的想法是让 Plasma 可以增加一些接口，使得能够做到 calendar 组件本身不需要感知到有“替代日历”的存在，然后通过这套接口来实现显示替代日历的功能，不过感觉工程量估计不小，而且不知道这么改的话比较大的 patch 是不是会给往上游推进造成困难，所以决定先把 plasma 的一部分代码搬出来试水（就是这个项目了）。因为时间不是很充裕所以完成度到够我自己用后也没继续搞下去 = =

稍远点说的话，说实话我挺希望 calendar 这块儿能直接拆出 plasma-framework 来维护，这样能做到方便的引入依赖而不影响 framework 这个仓库，也方便做功能细化，也方便各个其它组件直接使用 calendar 这个组件。不过方案选择上而言，如果能和 Kalendar 以及之前对 calendar decoration 支持有兴趣的人一起讨论下合适的引入方案的话，大概会好很多。

您的 MR 我早些时候看到相关动态的邮件后我大致扫过一眼，虽说没细看，大体上还是蛮不错的。目前您的实现来看，看上去替代日历功能与 calendar 小部件本身有一定程度耦合，我最初也想过这种写法（毕竟会省事一些），但不清楚那边能不能接受这样的方案。细节内容只大致看的话大概不够，目前也不是特别空闲，我这边有比较大片的空闲时间后重新细看下。

> 还需要对plasma-workspace的[数字时钟插件](https://github.com/KDE/plasma-workspace/blob/master/applets/digital-clock/package/contents/ui/CalendarView.qml)进行修改

所以官方要能把 calendar 从 plasma-workspace 里拆出来的话多好 (╯‵□′)╯︵┻━┻

> 目前SubLabel里的yearLabel和monthLabel还没有实际用途

我觉得对于农历而言月份和年份选择视图显示这个 label 似乎没很大作用，虽说可以显示壬寅年之类的东西，但这个信息如果别的地区的替代日历也用不到的话，（月和年的 subLabel）这个信息不如完全让插件放到别的地方呈现。不过我也不清楚这种想法是否合适。

> 不确定在Plasma里依赖ICU是否可行（虽然是可选的，且ICU一般发行版都有预装）

只提大致想法的话，我觉得 plasma framework 只需要能提供一个能在日期下方显示文字的 Label 的接口就可以了，然后供插件控制这个 label 的显示内容和显示优先级（我看您也有实现这个）。如果这样的话，ICU 的依赖也只会出现在额外的插件里而不会在 framework 里的 calendar 组件里，或许会更容易接受一些。

> ![image](https://user-images.githubusercontent.com/10095765/152829706-f8efaa1d-6b37-46b3-b31c-3fa18e691a85.png)

您这个示例截图的农历好像不太对的样子，1/31 应该是农历（腊月）**廿九**，2/1 是农历**正月**（初一），3/1 应该是农历（正月）**廿九**，不清楚您目前是否有修复这个问题。顺便，不清楚如果存在日历事件的情况下（日期格子有蓝点）日历会变成什么样子，如果也是在日期下方显示的话大概会显得整体靠下（我目前的实现也比较难看），这方面的问题不清楚 KDE 的 VDG 组会不会提供一些建议。

再次感谢，辛苦啦~

**Reply by easyteacher** [reply]
谢谢。SubLabel这个概念我觉得用处很广泛，节日、标签之类的都可以用到，至于上游不收或者想等Qt6移植完成后再做决定，那是另外一回事了。我的想法是先把概念提出来，这样上游能看见成品可能会方便决策一些。

关于蓝点问题，目前蓝点会显示在文字下方，需要重新设计一下。关于农历不正确的问题，可能是`d->day() == 1 ? d->monthDisplayName() : d->dayDisplayName()`这里并不需要用d->day()判断。其它细节部分还需要再研究一下。

**Reply by easyteacher** [reply]
关于农历问题，虽然不明白为什么直接在农历里设置日期不行，在Gregorian日历里设置日期然后从Gregorian日历里返回时间来设置农历就可以了。

**Reply by BLumia** [reply]
> SubLabel这个概念我觉得用处很广泛，节日、标签之类的都可以用到

+1，我的意思是月和年的 sublabel 不一定用得上，日期的是肯定需要的。如果月和年的 sublabel 在别的地区的日历系统有用的话大概可以保留吧。

> 关于农历问题，虽然不明白为什么直接在农历里设置日期不行，在Gregorian日历里设置日期然后从Gregorian日历里返回时间来设置农历就可以了。

直接设置日期（`set()`）设置的是对应日历的日期，在中国农历日历下设置的即为中国农历的日期而并非 Gregorian 日历的日期，应该只有 `setTime()` 是可以（即这个时间与日历无关）的。看上去您已经修复了 :)

另外目前看对应的几个 repo 还有很频繁的提交，不清楚是否还是 WIP 状态，等大致就绪后再细看好了~

**Reply by easyteacher** [reply]
目前还在调整细节，估计周末能弄完。

**Reply by easyteacher** [reply]
我自己也暂时看不出什么问题了，作者您有空的话可以来提几个审阅意见。谢谢😁

**Reply by BLumia** [reply]
好的，明天抽空看下，抱歉最近比较忙 😂

**Reply by BLumia** [reply]
比较仔细的看了目前您的实现，QML 我本身不是特别熟悉所以给不了评价，其它部分我感觉相当棒了。只看到一个小问题，已经评论在了对应的 MR 了 😄

**Reply by easyteacher** [reply]
谢谢作者！

**Comment by easyteacher** [comment]
在上游开了一个议题：https://invent.kde.org/frameworks/kdeclarative/-/issues/1

**Reply by BLumia** [reply]
在这个议题下回复了下我的简单看法，也希望之前在邮件列表里提到 calendar decoration 的开发者也可以参与讨论，以便最终可以提供足够通用的实现。

(p.s. 我之前的想法是 KHoliday 那一套给 event 增加一个 hint 性质/类别的 event，这种 event 不显示成蓝点而是把名称显示成 sublabel，但现在看来这种想法没有直接加 sublabel 通用的样子，也不能解决月和年没法有 sublabel 的问题)
(p.p.s. 比较仔细的看了目前您的实现，QML 我本身不是特别熟悉所以给不了评价，其它部分我感觉相当棒了)

> ![image](https://user-images.githubusercontent.com/10095765/153740834-c7a47a90-4818-4606-827c-e9eecee350b3.png) (引用自上面议题中 [@cblack](https://invent.kde.org/cblack) 的回复)

看来月和年的 sublabel 也是有用的 :joy:

**Comment by BLumia** [comment]
突然想到一个问题。（法定节假日）调休是国内仅有的做法吗？如果不是的话，是不是应该考虑提供适用于这种情况的 API？

**Reply by easyteacher** [reply]
看了下维基上关于美国节假日的介绍，可能只多放假，不调休。

> When a holiday occurs on Saturday or Sunday, that holiday is shifted to either Friday or Monday for work purposes.

**Reply by BLumia** [reply]
多放假的话应该也会有把工作日（周一到周五）标记为假期的需求？其实就是在考虑是否会需要把工作日标记为假期和把假期标记为工作日的需要，目前看上去主流（非国内的）日历应用都没提供这种功能，但至少对于国内来说这个还是有点必要的。

**Comment by easyteacher** [comment]
上周我在Plasma周会上提了一下日历的事情，然后现在的情况是有望在Plasma 5.26里并入上游，前提是[把plasma-framework里的calendar组件移动到plasma-workspace里](https://invent.kde.org/plasma/plasma-workspace/-/issues/41)（我记得作者提过这个，算是未卜先知了👍🏻），这段时间我争取持续跟进（催促

**Reply by BLumia** [reply]
移到 workspace 里也算拆出来维护了吧😂不过不用等 Plasma 6.x 就能合入还是挺好的消息。

好奇一下，周会有地方可以围观吗（（（

**Reply by easyteacher** [reply]
> 好奇一下，周会有地方可以围观吗（（（

周会是公开的，就在plasma聊天室里，理论上谁都可以发言的。

https://matrix.to/#/#plasma:kde.org

**Comment by easyteacher** [comment]
已经合并到5.26里了，后续就是添加节假日功能了。感谢作者！

**Reply by BLumia** [reply]
恭喜！
